### PR TITLE
Add staff grade details to "Your Grade" section

### PR DESCRIPTION
**Pull request by andy-armstrong**
## [TNL-3465](https://openedx.atlassian.net/browse/TNL-3465)

This PR updates the "Your Grade" section to the new design from the wireframes:

https://openedx.atlassian.net/wiki/display/TNL/%5BUX+Wireframes%5D+-+ORA+Staff+Grading

In particular, the staff grade is now presented for each criterion.

### Sandbox
- [x] https://andy-armstrong.sandbox.edx.org/courses/course-v1:edx+ORA203+course/info

### Testing
- [x] Data is properly encoded in HTML templates to avoid XSS
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @cahrens 
- [x] Code review: @efischer19 
- [ ] UI strings/error msgs review: @catong 
- [ ] UX review: @frrrances 
- [x] Product review: @explorerleslie 

FYI: @dianakhuang @scottrish 

### Post-review
- [x] Squash commits

### CODE REVIEW

**Comment by andy-armstrong** [comment]
@explorerleslie @frrrances @catong This feature is ready for testing on my sandbox:

https://andy-armstrong.sandbox.edx.org/courses/course-v1:edx+ORA203+course/info

The design doesn't exactly match the wireframes so please let me know if there are aspects that you think need improvement.

**Comment by andy-armstrong** [comment]
@cahrens @efischer19 I have all the tests passing here now, so I think this is ready for review. I'd still like to add some more test coverage, so let me know gaps that you see. In particular, I haven't added any staff assessments into the Python tests which is what I'm going to work on now.

**Comment by efischer19** [comment]
Looks good so far! I agree on more tests, especially around staff assessment/final grade interaction. Would also love to see this on a sandbox, if/when those issues are resolved.

**Comment by cahrens** [comment]
@andy-armstrong The grade is currently missing from within the table to indicate where the final grade came from.

image

File: openassessment/templates/openassessmentblock/grade/oa_assessment_title.html [hunk 1]
```diff
@@ -0,0 +1,35 @@
+{% load i18n %}
+{% spaceless %}
+    {% if assessment %}
+        <li class="answer">
+            <h5 class="answer__title">
+            <span class="answer__source">
+                {% if assessment.points > 0 %}
```
Comment (cahrens) [inline openassessment/templates/openassessmentblock/grade/oa_assessment_title.html]: Should also allow a score of 0-- this is the bug I showed you.

File: openassessment/templates/openassessmentblock/grade/oa_assessment_feedback.html [hunk 2]
```diff
@@ -0,0 +1,41 @@
+{% load i18n %}
+{% spaceless %}
+    {% if assessment %}
+        <li class="answer feedback">
+            {% if assessment.individual_assessments %}
+                {% for individual_assessment in assessment.individual_assessments %}
+                    {% if individual_assessment.feedback %}
```
Comment (cahrens) [inline openassessment/templates/openassessmentblock/grade/oa_assessment_feedback.html]: I don't understand this section of the template (lines 5-26). When is this shown?

File: openassessment/xblock/grade_mixin.py [hunk 3]
```diff
@@ -106,9 +110,9 @@ def render_grade_complete(self, workflow):
         if "peer-assessment" in assessment_steps:
             feedback = peer_api.get_assessment_feedback(submission_uuid)
             peer_assessments = [
-                self._assessment_grade_context(asmnt)
-                for asmnt in peer_api.get_assessments(submission_uuid)
-            ]
+                self._assessment_grade_context(peer_assessment)
+                for peer_assessment in peer_api.get_assessments(submission_uuid)
+                ]
```
Comment (cahrens) [inline openassessment/xblock/grade_mixin.py]: Whitespace error

File: openassessment/xblock/grade_mixin.py [hunk 4]
```diff
@@ -238,25 +226,23 @@ def submit_feedback(self, data, suffix=''):
...
-        Sanitize the rubric criteria into a format that can be passed
-        into the grade complete Django template.
-
-            * Add per-criterion feedback from peer assessments to the rubric criteria.
-            * Filters out empty feedback.
-            * Assign a "label" for criteria/options if none is defined (backwards compatibility).
+        Returns details about the grade assigned to the submission.
 
         Args:
             peer_assessments (list of dict): Serialized assessment models from the peer API.
```
Comment (cahrens) [inline openassessment/xblock/grade_mixin.py]: Missing some of the args.

File: openassessment/xblock/grade_mixin.py [hunk 5]
```diff
@@ -266,32 +252,189 @@ def _rubric_criteria_grade_context(self, peer_assessments, self_assessment):
...
+            'option': self._peer_median_option(submission_uuid, criterion),
+            'individual_assessments': [
+                _get_assessment_part(
+                    _('Peer {peer_index}').format(peer_index=index + 1),
+                    criterion_name,
+                    peer_assessment
+                )
+                for index, peer_assessment in enumerate(peer_assessments)
+                ],
+        } if len(peer_assessments) > 0 else None
```
Comment (cahrens) [inline openassessment/xblock/grade_mixin.py]: Yikes, I can't read this code. Can you change this around so it is more readable?

File: openassessment/xblock/grade_mixin.py [hunk 6]
```diff
@@ -266,32 +252,189 @@ def _rubric_criteria_grade_context(self, peer_assessments, self_assessment):
...
+            'criterion': criterion,
+            'option': self._peer_median_option(submission_uuid, criterion),
+            'individual_assessments': [
+                _get_assessment_part(
+                    _('Peer {peer_index}').format(peer_index=index + 1),
+                    criterion_name,
+                    peer_assessment
+                )
+                for index, peer_assessment in enumerate(peer_assessments)
+                ],
```
Comment (cahrens) [inline openassessment/xblock/grade_mixin.py]: Whitespace issue.

File: openassessment/xblock/grade_mixin.py [hunk 7]
```diff
@@ -266,32 +252,189 @@ def _rubric_criteria_grade_context(self, peer_assessments, self_assessment):
...
+        } if len(peer_assessments) > 0 else None
+        example_based_assessment_part = _get_assessment_part(
+            _('Example-Based Grade'), criterion_name, example_based_assessment
+        )
+        self_assessment_part = _get_assessment_part(_('Your Self Assessment'), criterion_name, self_assessment)
+
+        # TODO: automate this...
```
Comment (cahrens) [inline openassessment/xblock/grade_mixin.py]: What about this TODO?

File: openassessment/xblock/grade_mixin.py [hunk 8]
```diff
@@ -266,32 +252,189 @@ def _rubric_criteria_grade_context(self, peer_assessments, self_assessment):
...
+        if peer_assessment_part:
+            assessments.append(peer_assessment_part)
+        if example_based_assessment_part:
+            assessments.append(example_based_assessment_part)
+        if self_assessment_part:
+            assessments.append(self_assessment_part)
+
+        # Include points only for the first assessment
+        if len(assessments) > 0:
+            option = assessments[0]['points'] = assessments[0]['option']
```
Comment (cahrens) [inline openassessment/xblock/grade_mixin.py]: My poor brain can't handle multiple equal signs on a single line. Can you split this up?

File: openassessment/xblock/grade_mixin.py [hunk 9]
```diff
@@ -266,32 +252,189 @@ def _rubric_criteria_grade_context(self, peer_assessments, self_assessment):
...
+            if option:
+                assessments[0]['points'] = option['points']
```
Comment (cahrens) [inline openassessment/xblock/grade_mixin.py]: Again, I'm not understanding this as there appears to have been an assignment to assessments[0]['points'] right above.

File: openassessment/xblock/grade_mixin.py [hunk 10]
```diff
@@ -266,32 +252,189 @@ def _rubric_criteria_grade_context(self, peer_assessments, self_assessment):
...
+            submission_uuid (str): The id for the submission.
+            criterion (dict): The criterion in question.
+
+        Returns:
+            The option for the median peer grade.
+
+        """
+        median_scores = peer_api.get_assessment_median_scores(submission_uuid)
+        median_score = median_scores.get(criterion['name'], None)
+        return next((option for option in criterion['options'] if option['points'] == median_score), None)
```
Comment (cahrens) [inline openassessment/xblock/grade_mixin.py]: What if the point value either doesn't match one of the textual representations, or matches more than one (like in the test course, where "Good" and "Excellent" each correspond to the same number of points)?

**Comment by cahrens** [comment]
I agree about more tests-- there's a fair number of uncovered lines in grade_mixin. I would advise working to cover those lines, in particular in your new code (if not related to exception handling).

I'm done with my first pass at review.

File: openassessment/templates/openassessmentblock/grade/oa_assessment_title.html (hunk 1 above)
Comment (andy-armstrong) [inline openassessment/templates/openassessmentblock/grade/oa_assessment_title.html]: Thanks! I've fixed this, and updated the acceptance tests to verify the score.

File: openassessment/xblock/grade_mixin.py (hunk 3 above)
Comment (andy-armstrong) [inline openassessment/xblock/grade_mixin.py]: Nice catch! I don't know why this didn't cause any issues. We really need to get PyLint/PEP8 running...

File: openassessment/xblock/grade_mixin.py (hunk 8 above)
Comment (andy-armstrong) [inline openassessment/xblock/grade_mixin.py]: That was accidental. I think it was a bad cut-and-paste...

Comment (andy-armstrong) [inline openassessment/xblock/grade_mixin.py]: I agree completely that multiple equal signs are a bad idea

File: openassessment/xblock/grade_mixin.py (hunk 10 above)
Comment (andy-armstrong) [inline openassessment/xblock/grade_mixin.py]: Currently it ends up showing None for any of these cases. @explorerleslie what would you expect to be shown if the median score isn't a single unique option (either there are two options with the same score, or the median is in between the scores for two options). The three solutions that I can think of are:
1. Never show a value for the median, and only show the points
2. Show no value for the median only if there is not a single option (and show the points)
2. Show a hybrid value, e.g. "Good / Excellent"

File: openassessment/xblock/grade_mixin.py (hunk 7 above)
Comment (andy-armstrong) [inline openassessment/xblock/grade_mixin.py]: I was originally thinking about dynamically generating the parts, but I'm not going to do that now. I've removed the TODO

File: openassessment/xblock/grade_mixin.py (hunk 10 above)
Comment (explorerleslie) [inline openassessment/xblock/grade_mixin.py]: I'd prefer number 3.

**Comment by andy-armstrong** [comment]
@explorerleslie @catong @frrrances This feature is ready on my sandbox and I hope to merge tomorrow. For a fully assessed problem, see this page logged in as staff:

https://andy-armstrong.sandbox.edx.org/courses/course-v1:edx+ORA203+course/courseware/a4dfec19cf9b4a6fb5b18be6ccd9cecc/4dbd99f691064633a9be0daadb805ce6/

**Comment by explorerleslie** [comment]
@andy-armstrong I thought that some of the fonts were different but after inspecting I don't think they are. :+1: from me.

**Comment by cahrens** [comment]
I am uncomfortable with the label "Staff override". When we do the full staff grade workflow, it isn't so much an override as the expected grading workflow step. Will the label change in that case? Why is it necessary to put "Staff override" at all given that we are already in the "Staff" column?

FYI @explorerleslie and @catong 

image

Another question-- is it intentional that a student can see the names of the students who assessed him? I thought was anonymous, but the name of the student is being shown in the comments section.

image

**Comment by cahrens** [comment]
Oh wait, perhaps those "labels" are just the actual comments that @andy-armstrong wrote. In which case everyone can ignore my last comment...

File: openassessment/xblock/test/base.py [hunk 11]
```diff
@@ -172,3 +201,138 @@ class XBlockHandlerTransactionTestCase(XBlockHandlerTestCaseMixin, TransactionCa
...
+            },
+            'overall_feedback': u'Good job!',
+        },
+    ]
+
+    SUBMISSION = (u'ՇﻉรՇ', u'รપ๒๓ٱรรٱѻก')
+
+    STEPS = ['peer', 'self']
+
+    def _create_submission_and_assessments(
```
Comment (cahrens) [inline openassessment/xblock/test/base.py]: Nit: remove the "private" underscore prefix is you didn't use it for submit_staff_assessment (or set_staff_access). The methods should be consistent (what about set_mock_workflow_info?).

**Comment by cahrens** [comment]
:+1: You still need to follow up on the peer_media_option TODO, and I had a question above about labels vs. comment text. But otherwise everything looks awesome!

Don't forget to review my commit.

**Comment by catong** [comment]
@andy-armstrong Are updates to the titles in the Course Staff Tools and Course Staff Info modals (to match the new button labels) planned in a separate story/PR?

**Comment by andy-armstrong** [comment]
@catong I don't think so. I didn't realize that there were inconsistencies. Why don't you enumerate them here and I'll make them as part of this story.

**Comment by andy-armstrong** [comment]
@cahrens You are right that those labels were just feedback strings that I chose so I could see where they came from.

**Comment by catong** [comment]
@andy-armstrong Here you go:
Clicking "Manage Individual Learners" launches "Course Staff Tools"' (see screenshot); change heading within the modal to "Manage Individual Learners" to match the button label.
Clicking "View Assignment Statistics" launches "Course Staff Information" (see screenshot); change heading within the modal to "View Assignment Statistics" to match the button label.
manageindividuallearners
viewassignmentstats

Another question re behavior that I saw as I was getting these screenshots - the buttons turn blue on first click, but are pink when selected. Is this normal? It's rather garish when each button is in a different state! I know @clrux had a PR that changed button colors to dark pink in the RC yesterday. Buttons apparently were mistakenly set to blue when they should have been pink. Perhaps this behavior is only in dev/sandbox?
buttoncolors

**Comment by andy-armstrong** [comment]
@catong Ah, that was missed in the change to rename those buttons. @dianakhuang can you update the panel titles to match their buttons (in a separate PR).

**Comment by catong** [comment]
Thanks @andy-armstrong. Any thoughts about the button color issue?

**Comment by andy-armstrong** [comment]
@catong This PR doesn't make any changes to the staff tools, so I'm not sure if this is related to other changes. Can you report it as a bug and we'll investigate it separately.

**Comment by catong** [comment]
@andy-armstrong Will do, thanks!

**Comment by andy-armstrong** [comment]
@cahrens @efischer19 I've finally figured out all the edge cases for median peer scores (I think!). Please re-review. Note that I've also updated my sandbox with these changes. @cahrens I'll review your commits after my 1-on-1 with Scott.

File: openassessment/templates/openassessmentblock/grade/oa_grade_complete.html [hunk 12]
```diff
@@ -249,89 +106,92 @@ <h3 class="submission__feedback__title ui-toggle-visibility__control">
...
+                                                           class="option__input feedback__overall__value"
+                                                           value="Some comments I received were inappropriate."
+                                                           aria-labelledby="feedback__overall__prompt"/>
+                                                    <label class="option__label" for="feedback__overall__value--inappropriate">{% trans "Some comments I received were inappropriate." %}</label>
+                                                </li>
+                                            </ol>
                                         </li>
-                                    </ul>
+                                        <li class="field field--textarea feedback__remarks" id="feedback__remarks">
+                                            <label for="feedback__remarks__value">{% trans "Please provide any feedback on the grade or comments that you received from your peers." %}</label>
```
Comment (catong) [inline openassessment/templates/openassessmentblock/grade/oa_grade_complete.html]: @andy-armstrong Please strike "Please" and "any":
"Provide feedback on the grade or comments that you received from your peers."

**Comment by catong** [comment]
@andy-armstrong  Sorry if this is something I missed in the numerous discussions, but is it by design that in Steps 1 and 2 ("Your Response" and "Assess Peers") the headings are blue, but in Steps 3 and 4 ("Assess Your Response" and "Staff Grade") the headings are black? It's not completion, I don't think, as all steps are complete. 
orasteps

**Comment by andy-armstrong** [comment]
@catong Good question! This styling difference is there already in the released version of ORA so isn't part of this PR, e.g. you can see it on my test course on Edge:

https://edge.edx.org/courses/course-v1:AndyA+AAORA+1/courseware/a4dfec19cf9b4a6fb5b18be6ccd9cecc/338a4affb58a45459629e0566291381e/

My recollection is that there is a difference between blue titles that indicate a section that can be expanded, and black ones that cannot. I think the blue is intended to look like a link, meaning an interactive part of the page. @explorerleslie or @talbs can probably give you a more authoritative answer.

**Comment by efischer19** [comment]
:+1: That median options logic is fancy, took me second to get my head wrapped around it and verified. Looks great though!

**Comment by catong** [comment]
@andy-armstrong Thanks - the expandable explanation makes sense, though obviously the blue vs black obviously didn't trigger the "it's a link and the others are not" reaction in me. Also, not sure how effective/appropriate it is to use only color to make that distinction, from an accessibility standpoint.

**Comment by cahrens** [comment]
A couple of requests for clarity, otherwise :+1:

**Comment by cahrens** [comment]
We need to come up with a general strategy for IDs in cases where they must be used. The patterns I followed are the same patterns in other templates with radio button groups. I think it makes sense to fix them all at once instead of coming up with a scheme for this particular template.

**Comment by andy-armstrong** [comment]
:+1: for your excellent acceptance and a11y tests @cahrens.

**Comment by andy-armstrong** [comment]
@catong @frrrances I've squashed and am going to merge this PR once I've verified the tests against the sandbox. Please let me know if you plan on doing any further review. Thanks!

**Comment by catong** [comment]
@andyarmstrong nothing else from me. :+1:

**Pull request merged**